homeassistant/components/fan/reproduce_state.py

```diff
@@ -72,18 +81,46 @@ async def _async_reproduce_state(
     if state.state == STATE_ON:
         # The fan should be on
         if cur_state.state != STATE_ON:
-            # Turn on the fan at first
-            service_calls[SERVICE_TURN_ON] = service_data
-
-        for attr, service in ATTRIBUTES.items():
+            # Turn on the fan with all the speed and modes settings.
+            # The `turn_on` method will figure out in which mode to
+            # turn the fan on.
+            data = service_data.copy()
+            for attr in SPEED_AND_MODES_ATTRIBUTES:
+                if attr in state.attributes and state.attributes[attr] is not None:
```

Comment:
Is the `check_attr_equal` check OK to run here too? or is it avoided intentionally? I think it may not be needed in this case, but it seems like if it doesn't hurt then we can combine all attribute filters together.